src/CacheClient.cpp

```diff
@@ -0,0 +1,198 @@
+#include "Hash.hpp"
+#include "DagData.hpp"
+#include "RunAction.hpp"
+#include "NodeResultPrinting.hpp"
+#include "StatCache.hpp"
+#include "Profiler.hpp"
+#include "Exec.hpp"
+#include "BuildQueue.hpp"
+#include "MakeDirectories.hpp"
+
+const char* kENV_REAPI_CACHE_CLIENT = "REAPI_CACHE_CLIENT";
+const char* kENV_CACHE_SERVER_ADDRESS = "CACHE_SERVER_ADDRESS";
+const char* kENV_BEE_CACHE_BEHAVIOUR = "BEE_CACHE_BEHAVIOUR";
+
+static int SlowCallback(void *user_data, const char* label)
+{
+    SlowCallbackData *data = (SlowCallbackData *)user_data;
+    MutexLock(data->queue_lock);
+    char buffer[1000];
+    snprintf(buffer,sizeof(buffer),"%s %s", data->node_data->m_Annotation.Get(), label);
+    int sendNextCallbackIn = PrintNodeInProgress(data->node_data, data->time_of_start, data->build_queue, buffer);
+    MutexUnlock(data->queue_lock);
+    return sendNextCallbackIn;
+}
+
+static int SlowCallback_CacheRead(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheRead]");
+}
+
+static int SlowCallback_CacheWrite(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheWrite]");
+}
+
+enum Operation
+{
+    kOperationRead,
+    kOperationWrite
+};
+
+static int AppendFileToCommandLine(int totalWritten, char* buffer, int bufferSize, const char* fileName)
+{
+    int remainingBudget = bufferSize - totalWritten;
+
+    //headsup: requiredSpace return value does _not_ include length of the 0 terminator.
+    int requiredSpace = snprintf(buffer+totalWritten, remainingBudget, " \"%s\" ", fileName);
+
+    if (requiredSpace >= remainingBudget)
+    {
+        Log(kError, "Building CacheClient string exceeded buffer length");
+        return 0;
+    }
+
+    return totalWritten + requiredSpace;
+}
+
+
+static uint32_t s_CacheClientFailureCount = 0;
+const uint32_t kMaxClientFailureCount = 5;
+
+static CacheResult::Enum Invoke_REAPI_Cache_Client(const HashDigest& digest, StatCache *stat_cache, const FrozenArray<FrozenFileAndHash>& outputFiles, ThreadState* thread_state, Operation operation, const Frozen::Dag* dag, const Frozen::DagNode* dagNode, Mutex* queue_lock, const char* ingredients_file)
+{
+    if (s_CacheClientFailureCount > kMaxClientFailureCount)
+        return CacheResult::DidNotTry;
+
+    ProfilerScope profiler_scope("InvokeCacheMe", thread_state->m_ProfilerThreadId, outputFiles[0].m_Filename);
+
+    const char* reapi_raw = getenv(kENV_REAPI_CACHE_CLIENT);
+    if (reapi_raw == nullptr)
+        Croak("%s not setup", kENV_REAPI_CACHE_CLIENT);
+
+    PathBuffer pathbuf;
+    PathInit(&pathbuf, reapi_raw);
+    char reapi[kMaxPathLength];
+    PathFormat(reapi, &pathbuf);
+
+    char buffer[5000];
+    int totalWritten = 0;
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, digest);
+
+    const char* cmd = operation == kOperationRead ? "down" : "up";
+    totalWritten += snprintf(buffer, sizeof(buffer), "%s -v %s %s00000000000000000000000000000002", reapi, cmd, digestString);
```

Comment:
I think it's worth leaving a comment about the `00000000000000000000000000000002`.

Reply:
Agreed. Where did this actually come from?

Reply:
REAPI cache servers use lowercase SHA256 hashes as cache keys (the hash type is configurable, but most implementations use SHA256), and we're using a smaller hash with this extra padding to reach the required length. I guess the 2 at the end is a sort of magic number that we can increment if we want to avoid getting cache hits from earlier versions of tundra.

Reply:
Added a comment